File: 286. Walls and Gates/wallsAndGates.js
const wallsAndGates = (rooms) => {
  const ROWS = rooms.length;
  const COLS = rooms[0].length;

  const q = [];
  const visited = new Set();

  const adRoom = (r, c) => {
    if (r < 0 || r === ROWS || c < 0 || c === COLS || visited.has(`${r}-${c}`) || rooms[r][c] === -1) {
      return;
    }
    visited.add(`${r}-${c}`);
    q.push([r, c]);
  };

  for (let r = 0; r < ROWS; r++) {
    for (let c = 0; c < COLS; c++) {
      if (rooms[r][c] === 0) {
        q.push([r, c]);
        visited.add(`${r}-${c}`);
      }
    }
  }

  let dist = 0;
  while (q.length) {
    const currentLength = q.length;
    for (let i = 0; i < currentLength; i++) {
      const [r, c] = q.shift();
      rooms[r][c] = dist;
      adRoom(r + 1, c);
      adRoom(r - 1, c);
      adRoom(r, c - 1);
      adRoom(r, c + 1);
    }
    dist++;
  }
};

const grid = [
  [2147483647, -1, 0, 2147483647],
  [2147483647, 2147483647, 2147483647, -1],
  [2147483647, -1, 2147483647, -1],
  [0, -1, 2147483647, 2147483647]
];

console.log('wallsAndGates: ', wallsAndGates(grid));

File: 286. Walls and Gates/wallsAndGates.py
from collections import deque
from typing import List

class Solution:
  def wallsAndGates(self, rooms: List[List[int]]) -> None:
    """
    Do not return anything, modify rooms in-place instead.
    """
    ROWS, COLS = len(rooms), len(rooms[0])
    visited = set()
    q = deque()

    def addRoom(r, c):
      if (r < 0 or r == ROWS or c < 0 or c == COLS or
        (r, c) in visited or rooms[r][c] == -1):
        return
      visited.add((r, c))
      q.append([r, c])

    for r in range(ROWS):
      for c in range(COLS):
        if rooms[r][c] == 0:
          q.append([r, c])
          visited.add((r, c))

    dist = 0
    while q:
      for i in range(len(q)):
        r, c = q.popleft()
        rooms[r][c] = dist
        addRoom(r, c + 1)
        addRoom(r, c - 1)
        addRoom(r + 1, c)
        addRoom(r - 1, c)
      dist += 1


grid = [
  [2147483647, -1, 0, 2147483647],
  [2147483647, 2147483647, 2147483647, -1],
  [2147483647, -1, 2147483647, -1],
  [0, -1, 2147483647, 2147483647]
]

solution = Solution()
print('wallsAndGates: ', solution.wallsAndGates(grid))
